### ISSUE TITLE
IDB transaction is automatically closed while streaming transactionInfo

### ISSUE THREAD

**Issue by Valentine1898** [penumbra-zone/web#495]
Issue is related to #481 

- If we only have 1 transaction, we won't get an error  
- If we have 2 or more transactions, we will get the error `Failed to execute 'continue' on 'IDBCursor': The transaction has finished `
-  transaction is always closed after the second record is transferred 
- `IterateAssetsMetadata()` tests do not cause an error 

https://github.com/jakearchibald/idb?tab=readme-ov-file#transaction-lifetime 
Most likely, there is an await call somewhere at the transport level that causes the IDB transaction to close prematurely

**Comment by grod220** [comment]
Running into this for `unclaimedSwaps` endpoint as well. We should create a test that replicates this error so when its fixed, we are defended from it coming back.

**Comment by grod220** [comment]
Yikes. It looks like our internal messaging is not compatible with indexdb async iterables due to its restrictions on using async in between calls. Take a look at the call stack on the second iteration:

<img width="1017" alt="Screenshot 2024-02-13 at 11 46 42 AM" src="https://github.com/penumbra-zone/web/assets/16624263/c50ff829-0cce-47dd-a709-32bd70494a41">

It appears an await gets added to pass the result on to the client, which kills the idb connection.

**Comment by grod220** [comment]
For example, in `unclaimed-swaps.ts` this fixes it:

```typescript
  const swaps: { swap: SwapRecord }[] = [];
  for await (const swap of indexedDb.iterateSwaps()) {
    if (swap.heightClaimed !== 0n) continue;
    swaps.push({ swap });
  }

  for (const swap of swaps) {
    yield swap;
  }
```

Which is sad, because this basically is what we had before. This indexeddb restriction is quite unfortunate. It prevents us from having a true end-to-end way of streaming database results. I wonder if changing our transport streaming to async iterables would help 🤔 . But doesn't seem to be worth it at this stage.

**Comment by turbocrime** [comment]
> Yikes. It looks like our internal messaging is not compatible with indexdb async iterables due to its restrictions on using async in between calls.
> [...]
> It appears an await gets added to pass the result on to the client, which kills the idb connection.

it is frustrating that we can't await while iterating idb - but i think there would be an effective boundary at chrome runtime transport. i could try and refactor to address this. removing the rethrow wrapper is the most direct step. we'd have to do the try-catch wrapper in every impl function directly but i think this wouldn't be too bad

**Comment by turbocrime** [comment]
it seems connectrpc uses await internally

**Comment by grod220** [comment]
It feels like we are at a limitation imposed by chrome messaging 😢 . I almost think we should revert to `getAll()` again...

**Comment by turbocrime** [comment]
what exactly are the limitations on awaiting idb cursors? i can't find documentation

**Comment by grod220** [comment]
https://github.com/jakearchibald/idb?tab=readme-ov-file#transaction-lifetime

**Comment by turbocrime** [comment]
more streams may actually be the solution - the stream compartmentalizes, buffers the input, and internally is written synchronously. we might be able to rewrite these methods as transformers and it might shield the tx from being "awaited"

**Comment by turbocrime** [comment]
fixed by #504
